test/kino/input_test.exs

```diff
@@ -51,22 +51,61 @@ defmodule Kino.InputTest do
     end
   end
 
-  describe "date/2" do
-    test "allows default value at the edge of allowed range" do
-      assert %{attrs: %{default: ~D/2023-02-01/}} =
-               Kino.Input.date("Length",
-                 min: ~D/2023-02-01/,
-                 max: ~D/2023-03-01/,
-                 default: ~D/2023-02-01/
-               )
-    end
-  end
-
   describe "file/2" do
     test "raises an error when :accept is an empty list" do
       assert_raise ArgumentError, "expected :accept to be a non-empty list, got: []", fn ->
         Kino.Input.file("File", accept: [])
       end
     end
   end
+
+  describe "date/2" do
+    @valid_inputs [
+      # Simpler cases with no ambiguity
+      {[], %{default: nil, min: nil, max: nil}},
+      {[default: ~D[2023-09-01]], %{default: ~D[2023-09-01], min: nil, max: nil}},
+      {[min: ~D[2023-09-01]], %{default: nil, min: ~D[2023-09-01], max: nil}},
+      {[max: ~D[2023-09-01]], %{default: nil, min: nil, max: ~D[2023-09-01]}},
+      {[default: ~D[2023-09-01], min: ~D[2023-08-01]],
+       %{default: ~D[2023-09-01], min: ~D[2023-08-01], max: nil}},
+      {[default: ~D[2023-09-01], max: ~D[2023-09-30]],
+       %{default: ~D[2023-09-01], min: nil, max: ~D[2023-09-30]}},
+      {[min: ~D[2023-09-01], max: ~D[2023-09-30]],
+       %{default: nil, min: ~D[2023-09-01], max: ~D[2023-09-30]}},
+      {[default: ~D[2023-09-01], min: ~D[2023-08-01], max: ~D[2023-09-30]],
+       %{default: ~D[2023-09-01], min: ~D[2023-08-01], max: ~D[2023-09-30]}},
+      # Min value equals max
+      {[min: ~D[2023-09-01], max: ~D[2023-09-01]],
+       %{default: nil, min: ~D[2023-09-01], max: ~D[2023-09-01]}},
+      # Default value equal min and/or max
+      {[default: ~D[2023-09-01], min: ~D[2023-09-01], max: ~D[2023-09-30]],
+       %{default: ~D[2023-09-01], min: ~D[2023-09-01], max: ~D[2023-09-30]}},
+      {[default: ~D[2023-09-30], min: ~D[2023-09-01], max: ~D[2023-09-30]],
+       %{default: ~D[2023-09-30], min: ~D[2023-09-01], max: ~D[2023-09-30]}},
+      {[default: ~D[2023-09-01], min: ~D[2023-09-01], max: ~D[2023-09-01]],
+       %{default: ~D[2023-09-01], min: ~D[2023-09-01], max: ~D[2023-09-01]}}
+    ]
+
+    Enum.each(@valid_inputs, fn {opts, expected_attrs} ->
+      test "creates a new date input with opts #{inspect(opts)}" do
+        assert %{attrs: unquote(Macro.escape(Map.merge(expected_attrs, %{type: :date})))} =
+                 Kino.Input.date("Date Input", unquote(Macro.escape(opts)))
+      end
+    end)
```

Comment:
There are cases where generating tests with macros makes sense, but it has its drawbacks (running a specific test, quickly identifying what failed, readability). In this case, I don't think we need the whole matrix of cases. We can test a couple edge cases with different messages, this could actually be a single test :)